### ISSUE TITLE
build/gen_help.sh does not work on macOS as well as Windows

### ISSUE THREAD

**Issue by sergeyklay** [skvadrik/re2c#325]
Hello,

When I call on macOs `cmake ... -DRE2C_REBUILD_DOCS=ON ...` build fails with the following message:

```
<here is the base path>/.build/src/msg/help.cc:3:1: error: expected expression
;
^
1 error generated.
make[2]: *** [CMakeFiles/re2c_objects_autogen.dir/src/msg/help.cc.o] Error 1
make[1]: *** [CMakeFiles/re2c_objects_autogen.dir/all] Error 2
make: *** [all] Error 2
Error: Process completed with exit code 2.
```

A quick look at the `help.cc` file showed that it was generated incorrectly:
```
$ cat .build/src/msg/help.cc

extern const char *help;
const char *help =
;
```

I'm pretty sure that this is due to the fact that `build/gen_help.sh` doesn't use cross platform `sed` (hello BSD sed). Some similar problem in Windows (with using `bash.exe`).

Since we are planning continuous integration using the major OSes, I believe that any solution to this issue should include testing the documentation

**Comment by sergeyklay** [comment]
IMO, since the generation of documentation already depends on the presence of Python on host system (`rst2man.py`), the obvious solution to this issue would be rewriting this logic in Python, thereby making it cross-platform. I would gladly tackle this task, but I need a clearer instruction on what this script does and what is expected.

**Comment by sergeyklay** [comment]
By the way, there are a few more small Windows issues:

https://github.com/skvadrik/re2c/blob/7a5892fd91a5d6dd9506013f8ec6a36edd59cdf1/cmake/Re2cGenDocs.cmake#L6-L8
and
https://github.com/skvadrik/re2c/blob/7a5892fd91a5d6dd9506013f8ec6a36edd59cdf1/cmake/Re2cGenDocs.cmake#L28-L30

There is no need for redirect output with `>` (which doesn't work on Windows) as this is how `rst2man.py` works:

```text
$ rst2man.py --help
Usage
=====
  rst2man.py [options] [<source> [<destination>]]

...
...
...
```

**Comment by skvadrik** [comment]
There are a few different problems here.

> I'm pretty sure that this is due to the fact that build/gen_help.sh doesn't use cross platform sed (hello BSD sed).

I can reproduce this on FreeBSD. The problem is caused by the use of `head` utility (passing a negative number, which is a non-POSIX extension: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/head.html). 

> There is no need for redirect output with > (which doesn't work on Windows) as this is how rst2man.py works

Good catch, and easy to fix!

> IMO, since the generation of documentation already depends on the presence of Python on host system (rst2man.py), the obvious solution to this issue would be rewriting this logic in Python, thereby making it cross-platform. I would gladly tackle this task, but I need a clearer instruction on what this script does and what is expected.

Generally I think rewriting scripts in python is a good idea, especially large and complex ones. But in this particular case genhelp.sh is simple enough to use shell, and the portability issues can be fixed. I'm not opposed to rewriting it, but I'd like to fix the issue with `head` first.

> Since we are planning continuous integration using the major OSes, I believe that any solution to this issue should include testing the documentation

Agreed, basic functionality like manpages should work.

**Comment by sergeyklay** [comment]
> Generally I think rewriting scripts in python is a good idea, especially large and complex ones. But in this particular case genhelp.sh is simple enough to use shell, and the portability issues can be fixed. I'm not opposed to rewriting it, but I'd like to fix the issue with `head` first.

Well, could you prepare a patch then? This issue is blocking yet another PR I'm currently working on.

**Comment by sergeyklay** [comment]
Also, I'm just curious what this line does:
https://github.com/skvadrik/re2c/blob/7a5892fd91a5d6dd9506013f8ec6a36edd59cdf1/build/gen_help.sh#L1-L2

Is it used intentionally to truncate a file or check that it's readable? If so, could you please make this more explicit for human readers by using `true` or `:` as a dummy command, e.g. `true > "$2"` or `: > "$2"`? AFAIK, there are no semantic problems with using `> "$2"` over `true > "$2"`.

**Comment by sergeyklay** [comment]
> but I'd like to fix the issue with `head` first.

Here is my version:

```diff
diff --git a/build/gen_help.sh b/build/gen_help.sh
index 7678a26b..ab1395d3 100755
--- a/build/gen_help.sh
+++ b/build/gen_help.sh
@@ -1,13 +1,24 @@
+#!/usr/bin/env sh

-> "$2"
-echo "extern const char *help;" >> "$2"
-echo "const char *help =" >> "$2"
-MANPAGER=cat MANWIDTH=100 man "$1" \
-    | tail -n +6 \
-    | head -n -1 \
-    | sed -E -e 's/\x1b\x5b[0-9]+m//g' \
-    | sed 's/\\x/\\\\x/g' \
-    | sed -E 's/"/\\"/g' \
-    | sed -E 's/(.*)/"\1\\n"/' \
-    >> "$2"
-echo ";" >> "$2"
+{
+    echo "extern const char *help;"
+    echo "const char *help ="
+
+    # Here is how it works:
+    #
+    # 1. Drop six lines from the top
+    # 2. Drop two lines from the bottom
+    # 3. Remove all occurrences matched by regular expression '\x1b\x5b[0-9]+m'
+    # 4. Escape \\x by \\\\x
+    # 5. Escape " by \\"
+    # 6. Wrap each line in double quotes by adding a newline character ("\n")
+    MANPAGER=cat MANWIDTH=100 man "$1"     \
+        | tail -n +6                       \
+        | sed '$d' | sed '$d'              \
+        | sed -E -e 's/\x1b\x5b[0-9]+m//g' \
+        | sed 's/\\x/\\\\x/g'              \
+        | sed -E 's/"/\\"/g'               \
+        | sed -E 's/(.*)/"\1\\n"/'
+
+    echo ";"
+} > "$2"
```

Question: What does this regular expression match `\x1b\x5b[0-9]+m`? I'd like add it to a human readable variable.

**Comment by codylico** [comment]
> Question: What does this regular expression match \x1b\x5b[0-9]+m? I'd like add it to a human readable variable.

That looks like an escape sequence for terminal output, usually for color text and such (in particular SELECT GRAPHIC RENDITION for VT100).

**Comment by sergeyklay** [comment]
Hmm.. Thank you. It seems we do not need find escape sequences in the output in case they are not there :)
```diff
- MANPAGER=cat MANWIDTH=100 man "$1"
+ TERM=dumb man MANPAGER=cat MANWIDTH=100 man "$1"
```

**Comment by skvadrik** [comment]
@sergeyklay I experimented with the script on FreeBSD and realized that the whole approach with dropping first `n` lines and last `m` lines does not work, as man output differs on different systems (in fact, even on Linux man output has changed and the script was truncating the output too much, which was fixed recently).

If this is blocking your work, feel free to commit the above patch (however, I don't think it will work on FreeBSD). I'll experiment a bit more to see if I can find a better approach. To make it portable, we'll have to search for the line starting with `USAGE`. At the end it is even more vague, there is a variable number of empty lines and possibly some nonempty ones with parentheses. And the output width is slightly different on FreeBSD and Linux (by one symbol perhaps), I don't think we can help it. 

@codylico Thanks, well spotted!

**Comment by sergeyklay** [comment]
> And the output width is slightly different on FreeBSD and Linux (by one symbol perhaps), I don't think we can help it.

Do you mean `MANWIDTH=100`? Frankly speaking, I still don't quite understand what the meaning of output width is and why we need to deal with it at all. Could you please elaborate on that?

**Comment by skvadrik** [comment]
@sergeyklay `MANWIDTH` is needed to make the output nicely formatted (line wrap at 100 characters), otherwise they are wrapped at screen width.

**Comment by skvadrik** [comment]
By the way, I came up with sed expression `sed "/^USAGE/,//!d" | sed '$d'` that cuts off everything before `USAGE` and, and also the last line. But I can't make it work on FreeBSD.

**Comment by skvadrik** [comment]
@sergeyklay Regarding the use of `true > ...` instead of `> ...`, sure, this makes more sense. I'm just used to `> ...` and write it out of (bad) habit.

**Comment by sergeyklay** [comment]
> `MANWIDTH` is needed to make the output nicely formatted (line wrap at 100 characters), otherwise they are wrapped at screen width.

I'm sorry, still unclear :-/ Why do we need make the output nicely formatted and where? It is all about code generation? We are talking about the code that cannot be manually edited?

**Comment by sergeyklay** [comment]
In fact, regarding `help.cc` file, I would like to discuss the ways in which we can receive it. Windows does not have a native 'man' and 'sed' commands. However, there is a workaround by installing Cygwin. So we have to check for 'man' and 'sed' presence on Windows when we use `-DRE2C_REBUILD_DOCS`. This is not a guarantee that we will find it. This makes the build system tied to a specific systems. Even on systems that are native to 'man', we are forced to write strange code and use workarounds to sort out with UNIX and Linux. In general I don't like this approach because of its limitations.

**Comment by skvadrik** [comment]
The way we generate help message is from the .rst sources (shared between the manpage, help message and online documentation) to an intermediate manpage, and from the intermediate manpage to a C++ literal string: doc/help.rst.in -> (configure) help.rst -> (rst2man) help.rst.1 -> (gen_help.sh) help.cc. If we can bypass the intermediate manpage and generate C++ source code from the .rst sources directly, than we could get rid of the man invocation.

**Comment by skvadrik** [comment]
The main idea is to generate all documentation from the same .rst sources. It helps a lot, previously we had all sorts of mismatch between help message, online docs and the manpages. Online docs are synced to a different branch `gh-pages-gen` and used to generate re2c.org.

**Comment by skvadrik** [comment]
For example, we can write a python script rst2cpp.py that will convert help.rst to help.cpp.

**Comment by sergeyklay** [comment]
> By the way, I came up with sed expression `sed "/^USAGE/,//!d" | sed '$d'` that cuts off everything before `USAGE` and, and also the last line. But I can't make it work on FreeBSD.

Possible solution (tested on Linux only):

```shell
#!/usr/bin/env sh

# USAGE word line
lbound="$(TERM=dumb MANPAGER=cat MANWIDTH=100 man "$1" | grep USAGE -B 100 | wc -l)"

{
    echo "extern const char *help;"
    echo "const char *help ="

    # Here is how it works:
    #
    # 1. Drop lines from the top up to USAGE word
    # 2. Drop two lines from the bottom
    # 4. Escape \x by \\x
    # 5. Escape " by \"
    # 6. Wrap each line in double quotes " and adding a newline character \n
    TERM=dumb MANPAGER=cat MANWIDTH=100 man "$1" \
        | sed -e "1,${lbound}d"                  \
        | sed '$d' | sed '$d'                    \
        | sed 's/\\x/\\\\x/g'                    \
        | sed -E 's/"/\\"/g'                     \
        | sed -E 's/(.*)/"\1\\n"/'

    echo ";"
} > "$2"
```

**Comment by sergeyklay** [comment]
Another possible solution (this keeps USAGE word):

```shell
#!/usr/bin/env sh

{
    echo "extern const char *help;"
    echo "const char *help ="

    # Here is how it works:
    #
    # 1. Drop lines from the top up to USAGE word
    # 2. Drop two lines from the bottom
    # 4. Escape \x by \\x
    # 5. Escape " by \"
    # 6. Wrap each line in double quotes " and adding a newline character \n
    TERM=dumb MANPAGER=cat MANWIDTH=100 man "$1" \
        | grep -A 100 "USAGE"                    \
        | sed '$d' | sed '$d'                    \
        | sed 's/\\x/\\\\x/g'                    \
        | sed -E 's/"/\\"/g'                     \
        | sed -E 's/(.*)/"\1\\n"/'

    echo ";"
} > "$2"
```

**Comment by sergeyklay** [comment]
Yet another reason why build for macOs fails:

```shell
$ man /Users/runner/work/re2c/re2c/.build/src/msg/help.cc.1

Error reading man page /Users/runner/work/re2c/re2c/.build/src/msg/help.cc.1
No manual entry for /Users/runner/work/re2c/re2c/.build/src/msg/help.cc.1
```

Of course `/Users/runner/work/re2c/re2c/.build/src/msg/help.cc.1` file is exists and it contains the correct content.

I expected 'man' will read the specified page, as 'man' will treat the given argument as a file if it finds a slash `/` in it. It is works on Linux, however it doesn't work on macOs. Curious if this works in FreeBSD as far macOs is BSD-like system. I faced with this issue using man, version 1.6g and 1.6f

**Comment by sergeyklay** [comment]
Hmm... re2c 1.3 installed on Ubuntu Linux via apt displays a cropped version of the help

```bash
$ re2c --help | head -n 10
              Show help message.

       -1 --single-pass
              Deprecated. Does nothing (single pass is the default now).

       -8 --utf-8
              Generate  a  lexer that reads input in UTF-8 encoding.  re2c as‐
              sumes that character range is 0 -- 0x10FFFF and  character  size
              is 1 byte.
```

it should be:

```bash
$ re2c --help | head -n 10
USAGE
       re2c  [OPTION...] INPUT [-o OUTPUT]

       re2go [OPTION...] INPUT [-o OUTPUT]

OPTIONS

       -? -h --help
              Show help message.

       -1 --single-pass
              Deprecated. Does nothing (single pass is the default now).

       -1 --single-pass
              Deprecated. Does nothing (single pass is the default now).

       -8 --utf-8
              Generate  a  lexer that reads input in UTF-8 encoding.  re2c as‐
              sumes that character range is 0 -- 0x10FFFF and  character  size
              is 1 byte.
```

**Comment by sergeyklay** [comment]
Yet another issue with BSD man (at least on CI):

```shell
$ man grep > "grep.0"
$ cat "grep.0"
```

![image](https://user-images.githubusercontent.com/1256298/96494439-06290500-124f-11eb-9ee8-2be434b7c49a.png)

Everything is OK with GNU man.  Of course `man "$1"` for re2c produces the same.

At the screenshot `man` attempts to do underlines and bold text and maybe some other things with backspaces, duplicate characters, escape sequences, etc. Tricks that might work if you print the man output on a dot matrix or other printer, or show it as text on a terminal.

The solution:
```shell
#!/usr/bin/env sh

{
    # We can safely export the following variables unless we source this file
    export TERM=dumb
    export MANPAGER=cat
    export MANWIDTH=100

    echo "extern const char *help;"
    echo "const char *help ="

    # Here is how it works:
    #
    # 1. 'col -b' removes backspaces, 'col -x' replaces tabs with spaces
    # 2. Drop lines from the top up to USAGE word
    # 3. Drop two lines from the bottom
    # 4. Escape \x by \\x
    # 5. Escape " by \"
    # 6. Wrap each line in double quotes " and adding a newline character \n
    man "$1"                        \
        | col -bx                   \
        | grep -A 100 "USAGE"       \
        | sed '$d' | sed '$d'       \
        | sed 's/\\x/\\\\x/g'       \
        | sed -E 's/"/\\"/g'        \
        | sed -E 's/(.*)/"\1\\n"/'

    echo ";"
} > "$2"
```

**Comment by sergeyklay** [comment]
@skvadrik To summarize my research I created a PR: https://github.com/skvadrik/re2c/pull/326. Could you please take a look

**Comment by skvadrik** [comment]
@sergeyklay Great investigation, thank you! So many weird errors on one page. :D

**Comment by skvadrik** [comment]
> Hmm... re2c 1.3 installed on Ubuntu Linux via apt displays a cropped version of the help

That's the error I mentioned in https://github.com/skvadrik/re2c/issues/325#issuecomment-711433285 that was fixed recently. The output of `man` used to have longer header, but then it got shorter and the script started to cut too much.

**Comment by skvadrik** [comment]
> Yet another reason why build for macOs fails

It is still failing after #325, right? I don't see it fixed.

**Comment by sergeyklay** [comment]
> > Yet another reason why build for macOs fails
> 
> It is still failing after #325, right? I don't see it fixed.

No, I fixed it with this patch: https://github.com/skvadrik/re2c/pull/326. I'll send a new PR soon.

**Comment by sergeyklay** [comment]
@skvadrik Since the patch is already in the master branch should we close this issue?

**BSD-like systems:**
I hope documentation generation is now fully supported for BSD-like systems.

**Windows systems:**
The patch doesn't add full Windows support. However, in case of `-DRE2C_REBUILD_DOCS=ON` and the absence of the `sed` and `man`  we'll show error message and will stop build. 

Notice, we still need something like rst2cpp in order to make the process truly cross-platform and do not rely on Cygwin.

**Comment by skvadrik** [comment]
Nice work! I opened https://github.com/skvadrik/re2c/issues/327 to track the remaining work.
